src/torch/tests/gradchecks.py

```diff
@@ -0,0 +1,200 @@
+            image_data = pet.ImageData(init_image_file)
+            acq_model = pet.AcquisitionModelUsingParallelproj()
+        else:
+            raise ValueError("Invalid data_type for PET. Choose '2d', '3d'.")
+    elif modality == "MR":
+        mr.AcquisitionData.set_storage_scheme('memory')
+        mr_data_path = examples_data_path('MR')
+        if data_type == "2d":
+            raw_data_file = existing_filepath(mr_data_path, 'simulated_MR_2D_cartesian.h5')
+            input_data = mr.AcquisitionData(raw_data_file)
+            prep_gadgets = ['RemoveROOversamplingGadget']
+            processed_data = input_data.process(prep_gadgets)
+            recon = mr.FullySampledReconstructor()
+            recon.set_input(processed_data)
+            recon.process()
+            complex_images = recon.get_output()
+            image_data = complex_images
+            csms = mr.CoilSensitivityData()
+            processed_data.sort()
+            csms.calculate(processed_data)
+            acq_model = mr.AcquisitionModel(processed_data, csms)
+            acq_model.set_coil_sensitivity_maps(csms)
+            acq_data = acq_model.forward(complex_images)
+        else:
+            raise ValueError("Invalid data_type for MR. Choose '2d'.")
+    else:
+        raise ValueError("Invalid modality. Only 'PET' or 'MR' are supported.")
+
+    return acq_data, image_data, acq_model
+
+
+@pytest.fixture(params=[
+    ("MR", "2d"), ("PET", "2d"), ("PET", "3d"), 
+])
+def test_data(request):
+    modality, data_type = request.param
+    acq_data, image_data, acq_model = get_data(modality, data_type)
+    image_data = image_data.get_uniform_copy(1.)
+    acq_model.set_up(acq_data, image_data)
+    acq_data = acq_model.forward(image_data)
+    return acq_data, image_data, acq_model, modality, data_type
+
+
+test_flags = {
+    "forward": True,
+    "adjoint": True,
+    "objective_wrapped": True,
+    "objective": True,
+    "objective_gradient": True,
+}
+
+def run_gradcheck(func, input_data, data_info, test_name, **kwargs):
+    """Helper function to run gradcheck and handle exceptions."""
+    modality, data_type = data_info
+    try:
+        torch.autograd.gradcheck(func, input_data, raise_exception=True, **kwargs)
+        print(f"{test_name} passed for {modality} {data_type}")
+        assert True
+    except Exception as e:
+        print(f"{test_name} failed for {modality} {data_type}")
+        print(e)
+        assert False
+
+
+@pytest.mark.skipif(not test_flags["forward"], reason="Forward test disabled")
+def test_forward_gradcheck(test_data):
+    acq_data, image_data, acq_model, modality, data_type = test_data
+    if modality == "PET":
+        pass
+    elif modality == "MR":
+        pass
+    else:
+        pytest.skip("Tests not set up for other modalities at this time.")
+
+    torch_forward = Operator(acq_model, image_data.clone())
+    torch_image_data = torch.tensor(image_data.as_array(), requires_grad=True).unsqueeze(0).cuda()
```

Comment:
could we use `sirf_to_torch` here? (can the `unsqueeze` be run on its return?). Same comment applies in other test functions.

We should use a default device along the lines of
```
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
such that this can be run even if no GPU available. Not 100% sure how to do this with `pytest`, but I guess simplest would be to have that line in every test function.

An alternative would be to have `sirf_to_torch` have a `default` option for `device`  accordingly.